File: c#/bitcoin/Number/Number2.cs
using System;
using System.Numerics;

public class Number2 : Number 
{
  // data
  private readonly BigInteger val;

  // constructor
  public Number2(BigInteger val){ this.val = val; }

  // Unlike java, static members in C# are inherited from the base class.
  // Hence the below declarations hides existing static members of class 
  // 'Number' which generates a compiler warning. The keyword 'new' is
  // required to suppress this warning.

  public new static Number ZERO = new Number2(BigInteger.Zero);

  public new static Number ONE = new Number2(BigInteger.One);

  public override Number Add(Number rhs)
  {
    // java NullPointerException
    if(rhs == null) throw new NullReferenceException("rhs is null");

    // java getClass()
    if(this.GetType() != rhs.GetType())
    {
      // java ClassCastException
      throw new InvalidCastException("rhs has illegal type");
    }

    Number2 cast = (Number2) rhs;

    return new Number2(BigInteger.Add(this.val, cast.val));
  }

  public override Number Mul(Number rhs)
  {
    // java NullPointerException
    if(rhs == null) throw new NullReferenceException("rhs is null");

    // java getClass()
    if(this.GetType() != rhs.GetType())
    {
      // java ClassCastException
      throw new InvalidCastException("rhs has illegal type");
    }

    Number2 cast = (Number2) rhs;

    return new Number2(BigInteger.Multiply(val, cast.val));
}
  
  public override Number Negate()
  {
    return new Number2(BigInteger.Negate(val));
  }

  public override string ToString()
  {
    return val.ToString();
  }

  public override int CompareTo(Number rhs)
  {

    // java NullPointerException
    if(rhs == null) throw new NullReferenceException("rhs is null");

    // java getClass()
    if(this.GetType() != rhs.GetType())
    {
      // java ClassCastException
      throw new InvalidCastException("rhs has illegal type");
    }

    Number2 cast = (Number2) rhs;

    return val.CompareTo(cast.val);
  }

  public override int GetHashCode()
  {
    return GetType().GetHashCode() + val.GetHashCode();
  }

  public override int Sign
  {
    get { return val.Sign; }
  }

  public override byte[] ToBytes(int numBytes)
  {
    byte[] unsigned = null; // returned byte array

    int sign = val.Sign;

    switch(sign)
    {
      case 0:

        unsigned = new byte[numBytes];

        for(int i = 0; i < numBytes; ++i)
        {
          unsigned[i] = (byte) 0x00;
        }

        break;

      case 1:

        // Unlike java's 'toByteArray', C# 'ToByteArray' returns a little
        // endian encoding of the number. Both java and C# return a two's 
        // complement representation.
        
        byte[] signedLittleEndian = val.ToByteArray();

        int len = signedLittleEndian.Length;

        byte[] signed = new byte[len];  // big-endian

        int j = len -1 ;

        for(int i = 0; i < len; ++i)
        {
          signed[i] = signedLittleEndian[j]; 
          
          --j;
        }

        unsigned = new byte[numBytes];
        
        // the signed array may have a leading 0x00 byte to ensure the number
        // is not wrongly interpreted as a negative number. If this is the case
        // then it would be wrong to assume the length of the signed array is
        // the amount of bytes required to encode the magnitude of the number.
        
        int size;   // number of bytes needed to encode magnitude of number
        int start;  // position from which signed array should copied over

        // note that sign = 1 ensures that signed cannot be the empty array
        
        if((signed[0] & 0xFF) == 0) // signed array has leading 0x00 byte
        {
          size = signed.Length - 1; // required size is one less
          start = 1;                // copy signed array starting from index 1
        }
        else
        {
          size = signed.Length;
          start = 0;
        }

        if(numBytes < size)
        {
          throw new ArithmeticException("numBytes argument is too small");
        }

        // number of leading 0x00 bytes which need to be padded

        int padding = numBytes - size;

        for(int i = 0; i < padding; ++i)
        {
          unsigned[i] = (byte) 0x00;
        }

        // copying signed array over, starting from the correct position 'start'

        Array.Copy(signed, start, unsigned, padding, size);

        break;

      case -1:

        unsigned = Negate().ToBytes(numBytes);  // recursive call

        break;

      default:  // should not happen

        // java also ArithmeticException
        throw new ArithmeticException("Unexpected error"); 
    }

    return unsigned;
  }

  public override int BitLength()
  {
    if(val.Sign >= 0)
    {
      byte[] bytes = val.ToByteArray();   // little-endian, two's complement

      int len = bytes.Length;

      if(len == 0) return 0;              // number can only be zero

      if((bytes[len - 1] & 0xFF) == 0)    // leading byte is 0x00 to indicate >=0
      {
        --len;
      }

      if(len == 0) return 0;

      int length = 8 * len;               // first estimate

      byte lead = bytes[len - 1];

      int mask = 0x80;                    // 1000 0000 in binary

      while((lead & mask) == 0)           // leading bit is actually 0
      {
        --length;

        mask = mask >> 1;                 // testing next bit to the right
      }

      return length;

    }
    else
    {
      return Negate().BitLength();
    }
  }

  public override BigInteger ToBigInteger()
  {
    return val;
  }
}

File: haskell/bitcoin/Number/Number2.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Number2 
  ( Number2
  , NumBytes(..)
  , NumBits(..)
  , Sign(..)
  , zero
  , one
  , fromBytes
  , random
  , toBytes
  , bitLength
  , sign
  , hash
  ) where

import INumber
import Rand
import Data.Hashable (Hashable, hash)
import Data.Word (Word8)
import Data.Bits (shiftR, shiftL, testBit, (.&.))
import Data.ByteString hiding (putStrLn, length)
import Prelude hiding (all, head, tail)


newtype Number2 = Number2 Integer 
        deriving (Eq, Ord, Num, Real, Enum, Integral, Hashable) 

instance INumber Number2 where
  zero      = zero_
  one       = one_
  fromBytes = fromBytes_
  random    = random_
  toBytes   = toBytes_
  sign      = sign_
  bitLength = bitLength_

instance Show Number2 where 
  show      = show_

show_ :: Number2 -> String
show_ (Number2 x) = (show x)

zero_ :: Number2
zero_  = Number2 0 

one_  :: Number2
one_   = Number2 1

sign_ :: Number2 -> Int
sign_ (Number2 x)
  | x == 0  = 0
  | x <= 0  = -1
  | x >= 0  = 1


fromBytes_ :: Sign -> ByteString -> Rand Number2  -- big endian
fromBytes_  (Sign sig) bytes
  | sig == 1    = return $ Number2 x
  | sig == (-1) = return  $ Number2 (-x) 
  | sig == 0    = 
      if all (== 0) bytes
        then return $ Number2 0
        else throw $
          randException "InvalidArgument" "fromBytes: inconsistent arguments"
  | otherwise   = throw $ 
    randException "InvalidArgument" "fromBytes: illegal sign argument"
  where 
    x     = go 0 list
    list  = unpack bytes
    go :: Integer -> [Word8] -> Integer 
    go acc [] = acc
    go acc (b:bs) = go ((shiftL acc 8) + toInteger b) bs 

-- returns unsigned random number of requested size in bits
random_ :: NumBits -> Rand Number2
random_ n = do
  bytes <- randomBytes_ n 
  fromBytes (Sign 1) bytes

-- returns unsigned random number as big endian ByteArray
-- Essentially generates random bytes and subsequently set
-- the appropriate number of leading bits to 0 so as to 
-- ensure the final ByteArray has the right bit size.
randomBytes_ :: NumBits -> Rand ByteString
randomBytes_ (NumBits n) = 
  let len = div (n + 7) 8 in    -- number of bytes required
    if len == 0 
      then return empty         -- empty ByteString
      else do
        bytes <- getRandomBytes len
        let lead = head bytes
        let diff = len * 8 - n  -- number of leading bits set to 0 
        let front = truncate_ diff lead
        return $ cons front (tail bytes)

-- returns byte with n leading bits set to 0
truncate_ :: Int -> Word8 -> Word8  
truncate_ n w = go 0x7f n w where
  go :: Word8 -> Int -> Word8 -> Word8
  go mask n w = 
    if n == 0
      then w
      else go (mask `shiftR` 1) (n - 1) (w .&. mask)
    

testTruncate :: IO ()
testTruncate = do
  putStrLn $ show (truncate_ 0 0xff)
  putStrLn $ show (truncate_ 1 0xff)
  putStrLn $ show (truncate_ 2 0xff)
  putStrLn $ show (truncate_ 3 0xff)
  putStrLn $ show (truncate_ 4 0xff)
  putStrLn $ show (truncate_ 5 0xff)
  putStrLn $ show (truncate_ 6 0xff)
  putStrLn $ show (truncate_ 7 0xff)
  putStrLn $ show (truncate_ 8 0xff)
     

toBytes_ :: Number2 -> NumBytes -> Rand ByteString
toBytes_  (Number2 x) (NumBytes num)
  | x < 0     = toBytes (Number2 $ -x) (NumBytes num)
  | len < 0   = throw $ 
    randException "InvalidArgument" "toBytes: NumBytes argument is too small"
  | otherwise = return $ pack $ padding len list 
  where
    list = (unroll x)
    len = num - length list


-- number of bits required to encode magnitude of number
bitLength_ :: Number2 -> NumBits
bitLength_  (Number2 x)
  | x < 0     = bitLength (Number2 $ -x)
  | len == 0  = NumBits 0
  | otherwise = NumBits (len * 8 - adjust)
  where
    list    = unroll x
    len     = length list
    byte    = list !! 0
    adjust  = leadBitCount byte
  

-- returns the number of leading bits equal to 0
leadBitCount :: Word8 -> Int
leadBitCount byte = go 0 7 byte where
  go count position byte 
    | position < 0          = count
    | testBit byte position = count
    | otherwise             = go (count + 1) (position - 1) byte 
    

-- adding a certain number of leading 0x00 bytes
padding :: Int -> [Word8] -> [Word8]
padding num list
  | num <= 0    = list
  | otherwise   = padding (num - 1) ((0x00):list)

-- applying unfold to convert Integer to list of bytes in big-endian format
unroll :: Integer -> [Word8]
unroll = unfold step where
  step 0 = Nothing
  step i = Just (fromInteger i, i `shiftR` 8) 

-- our own version of unfold, so as to end up with big-endian 
unfold :: (b -> Maybe (a,b)) -> b -> [a]
unfold step b = go b [] where
  go b acc =
    case step b of
      Nothing     -> acc
      Just (a,b') -> go b' (a:acc) -- tail recursive


